whatweb

```diff
@@ -366,28 +329,27 @@ def open_target( url )
 end
 
 ### output
-
-def error( s )
-	return if $NO_ERRORS
-	if defined?( $semaphore )
-		# We want the output mutex locked.
-		# Has our current thread already locked the Mutex?
-		begin
-			$semaphore.lock
-		rescue ThreadError
-			# we're already locked. This was expected.
-		end
-	end
-	if ( $use_colour=="auto" ) or ( $use_colour=="always" )
-		STDERR.puts red( s )
-	else
-		STDERR.puts s
-	end
-	STDERR.flush
-	unless $OUTPUT_ERRORS.nil?
-		$OUTPUT_ERRORS.out( s )
-	end
-	$semaphore.unlock if defined?( $semaphore )
+$semaphore = Mutex.new
+# this is noisy in debug
+def $semaphore.reentrant_synchronize(&work)
```

Comment:
In the current incarnation, this is only used once and is a bit defensive. If the use of `$semaphore` is dropped from lib/output.rb, this method could be dropped and the single use replaced with a plain `.synchronize` - then `$semaphore` would exclusively be used for error logging.